lib/src/compiler/parse.rs

```diff
@@ -1,14 +1,220 @@
+        alt((
+            parse_read,
+            parse_write,
+            parse_set,
+            parse_push,
+            parse_pop,
+            parse_if,
+            parse_while,
+        )),
+    )(input)?;
+    Ok((input, res))
+}
+
+fn parse_body(input: &str) -> IResult<&str, Vec<Instr>> {
+    let (input, res) = delimited(
+        preceded(multispace0, char('{')),
+        many0(try_each),
+        preceded(multispace0, char('}')),
+    )(input)?;
+    Ok((input, res))
+}
+
+fn my_parse(input: &str) -> IResult<&str, Vec<Instr>> {
+    let (input, res) = terminated(many1(try_each), multispace0)(input)?;
+    Ok((input, res))
+}
+
 impl Compiler<()> {
     /// Parses source code from the given input, into an abstract syntax tree.
     pub fn parse(self, source: &mut impl Read) -> Fallible<Compiler<Program>> {
         let mut source_buffer = String::new();
         source.read_to_string(&mut source_buffer)?;
-        // TODO real parsing here
-        let program = serde_json::from_str(&source_buffer)?;
-        Ok(Compiler(program))
+        match my_parse(&source_buffer) {
+            // TODO: can probably refactor the parser funcs to use
+            // Verbose error to make the errors nicer
+            // example: https://github.com/Geal/nom/blob/master/examples/s_expression.rs
+            Ok((_, body)) => {
+                let prog = Program { body };
+                Ok(Compiler(prog))
+            }
+            Err(nom::Err::Error((_input, e))) => {
+                Err(ParseError::Generic(e.description().to_string()).into())
+            }
+            Err(nom::Err::Incomplete(_needed)) => {
+                // TODO: better ass
+                Err(ParseError::Generic("ass".to_string()).into())
+            }
+            Err(nom::Err::Failure((_input, e))) => {
+                Err(ParseError::Generic(e.description().to_string()).into())
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    #[test]
+    fn test_parse_read_write() {
+        assert_eq!(
+            my_parse(
+                "
+            Read
+            Write
+        "
+            ),
+            Ok(("", vec![Instr::Read, Instr::Write]))
+        )
+    }
+    #[test]
+    fn test_set() {
+        assert_eq!(my_parse("Set 4"), Ok(("", vec![Instr::Set(4),])))
+    }
+
+    #[test]
+    fn test_push() {
+        assert_eq!(my_parse("Push 4"), Ok(("", vec![Instr::Push(4),])))
+    }
+
+    #[test]
+    fn test_pop() {
+        assert_eq!(my_parse("Pop 4"), Ok(("", vec![Instr::Pop(4),])))
+    }
+
+    #[test]
+    fn test_parse_if() {
+        assert_eq!(
+            my_parse(
+                "if {
+            Read
+            Write
+        }"
+            ),
+            Ok(("", vec![Instr::If(vec![Instr::Read, Instr::Write,])]))
+        )
+    }
+
+    #[test]
+    fn test_parse_while() {
+        assert_eq!(
+            my_parse(
+                "while {
+            Read
+            Write
+        }"
```

Comment:
I'd be down to support either any casing, so you could write `WHILE {...}` or `while {...}` or whatever, or just all caps, since that feels old and memey.

Reply:
making it not care about case. Maybe we should require cases like `MeMeS`

Reply:
spongebob case? that'd be dank lul